salt/returners/prometheus_textfile.py

```diff
@@ -0,0 +1,332 @@
+    duration = 0
+    for state, data in ret.get("return", {}).items():
+        total += 1
+        duration += data.get("duration", 0)
+        if data["result"]:
+            success += 1
+        else:
+            failure += 1
+        if data.get("changes"):
+            changed += 1
+
+    if not total:
+        log.error("Total states run equals 0. There may be something wrong...")
+        raise  # pylint: disable=misplaced-bare-raise
+
+    salt_procs = 0
+    if HAS_PSUTIL:
+        salt_procs = len(
+            _find_minion_procs(
+                proc_name=opts["proc_name"],
+                match_exe=opts["match_exe"],
+                exe=opts["exe"],
+            )
+        )
+
+    now = int(time.time())
+
+    output = {
+        "salt_procs": {
+            "help": "Number of salt minion processes running",
+            "value": salt_procs,
+        },
+        "salt_states_succeeded": {
+            "help": "Number of successful states in the run",
+            "value": success,
+        },
+        "salt_states_failed": {
+            "help": "Number of failed states in the run",
+            "value": failure,
+        },
+        "salt_states_changed": {
+            "help": "Number of changed states in the run",
+            "value": changed,
+        },
+        "salt_states_total": {
+            "help": "Total states in the run",
+            "value": total,
+        },
+        "salt_states_success_pct": {
+            "help": "Percent of successful states in the run",
+            "value": round((success / total) * 100, 2),
+        },
+        "salt_states_failure_pct": {
+            "help": "Percent of failed states in the run",
+            "value": round((failure / total) * 100, 2),
+        },
+        "salt_states_changed_pct": {
+            "help": "Percent of changed states in the run",
+            "value": round((changed / total) * 100, 2),
+        },
+        "salt_elapsed_time": {
+            "help": "Time spent for all operations during the state run",
+            "value": round(duration, 3),
+        },
+        "salt_last_started": {
+            "help": "Estimated time the state run started",
+            "value": int(now - duration / 1000),
+        },
+        "salt_last_completed": {
+            "help": "Time of last state run completion",
+            "value": now,
+        },
+    }
+
+    if opts["add_state_name"]:
+        old_name = os.path.basename(opts["filename"])
+        new_name = (
+            "".join(old_name.split(".")[:-1])
+            + "-{}.".format(prom_state)
+            + "".join(old_name.split(".")[-1])
+        )
+        opts["filename"] = os.path.join(out_dir, new_name)
+        log.debug("Modified Prometheus filename from %s to %s", old_name, new_name)
+        for key in list(output.keys()):
+            output[key + '{state="' + prom_state + '"}'] = output.pop(key)
+
+    if opts["mode"]:
+        try:
+            opts["mode"] = int(opts["mode"], base=8)
```

Comment:
**question** given the mode validation, should we also validate that `0 < int(opt["mode"], 8) < 511` (aka `0o777`,  or `stat.S_IWUSR | stat.S_IXUSR | stat.S_IRUSR | stat.S_IWGRP | stat.S_IXGRP | stat.S_IRGRP | stat.S_IWOTH | stat.S_IROTH | stat.S_IXOTH`)?